### ISSUE TITLE
"This module requires the OpenShift Python client. Try `pip install openshift`"

### ISSUE THREAD

**Issue by yusuke0324** [NVIDIA/deepops#321]
Hi,

I ran `$ansible-playbook -i k8s-config/hosts.ini -b playbooks/k8s-cluster.ml -u deepops -K` with ansible 2.7.11 .on Ubuntu 16.04. (using 19.03 and python 2.7) and got the following error.
~~~
TASK [create GPU device plugin] **************************************************************************************
fatal: [node1]: FAILED! => {"changed": false, "msg": "This module requires the OpenShift Python client. Try `pip install openshift`"}
~~~

I did `pip install openshift` on my venv where I ran the above command but it made no change.

**Comment by dholt** [comment]
Hi, we don't officially support or test on Ubuntu 16.04.

To work around it, you'll need the openshift python client on your target machine, not your provisioning machine. The role installs and uses a python virtual env in `/opt/deepops` by default. See: `roles/k8s-gpu-plugin/tasks/main.yml`.

**Comment by yusuke0324** [comment]
I did `$sudo pip install openshift` on the target machine but got the same error.
Could you give me a bit more instruction about the workaround?

**Comment by dholt** [comment]
Try `sudo -i`, `source /opt/deepops/bin/activate`, then `pip install openshift`

**Comment by yusuke0324** [comment]
I might be missing something but do I need to git clone deepops at `/opt/deepos/`? Because I cannot find `/opt/deepops/`.
~~~
root@node1:/opt# source /opt/deepops/bin/activate
-bash: /opt/deepops/bin/activate: No such file or directory
~~~

The target machine is not the machine I run `ansible-playbook`, correct?

**Comment by yusuke0324** [comment]
seems like `roles/k8s-gpu-plugin/tasks/main.yml` exists in `19.07` but not in `19.03` version. I'm trying to set up with `19.03` due to Ubuntu 16.04. Are `/opt/deepops/` and  `virtual env` created with `19.07` too?

**Comment by ajdecon** [comment]
Short version: this appears to be a bug that was fixed in the 19.07 release.

The directory @dholt refers to was created as part of this commit that changed how we install the GPU plugin:

https://github.com/NVIDIA/deepops/commit/7a3315cdf2ab1ccaae5f4cba4559e92e76c4c74d

This commit was added after the 19.03 release, which explains where you aren’t seeing the directory here.

IIRC that commit was added due to some conflicting Python package versions on Ubuntu. We started using a virtualenv because we sometimes saw issues like this where the k8s ansible module couldn’t consistently find the openshift pip package, even if it was installed by the system Python.

You may be able to work around this issue by   commenting out the k8s device plugin install in  the k8s-cluster.yml playbook, and then installing that plugin manually using the instructions here: https://github.com/NVIDIA/k8s-device-plugin

But as @dholt says, the right answer here is that Ubuntu 16.04 is not officially supported, and we can’t guarantee that it will continue to work, especially where it pulls in external dependencies.
________________________________
